Internal error during connect: module "toolhead" has no attribute "BUFFER_TIME_HIGH"

as the title suggests i get this error when including the various mmu.cfg

I'm stuck in other place now, after applying the patch BUFFER_TIME_HIGH error disappeared but other error appears:

`[INFO] 2024-09-04 20:22:48,718 [root] [mcu:_mcu_identify:823] Loaded MCU 'mmu' 120 commands (?-20240824_231541-539011865e59 / gcc
: (15:12.2.rel1-1) 12.2.1 20221205 binutils: (2.40-2+18+b1) 2.40)
MCU 'mmu' config: ADC_MAX=4095 CLOCK_FREQ=48000000 MCU=samd21g18a PWM_MAX=255 RESERVE_PINS_USB=PA24,PA25 STATS_SUMSQ_BASE=256 STE
PPER_BOTH_EDGE=1
[INFO] 2024-09-04 20:22:48,720 [root] [serialhdl:connect_pipe:167] mcu 'rpi': Starting connect
[INFO] 2024-09-04 20:22:48,725 [root] [serialhdl:_bg_thread:40] mcu 'rpi': current nice = -20
[INFO] 2024-09-04 20:22:49,411 [root] [mcu:_mcu_identify:823] Loaded MCU 'rpi' 104 commands (v0.11.0-372-gb9ad7605 / gcc: (Ingeni
c Linux-Release5.1.0-Default(xburst2(fp64)+glibc2.29) 2021.12-22 10:52:10) 7.2.0 binutils: (Ingenic Linux-Release5.1.0.-Default(x
burst2(fp64)+glibc2.29) 2021.12-22 08:55:37) 2.27)
MCU 'rpi' config: ADC_MAX=4095 CLOCK_FREQ=50000000 MCU=linux PCA9685_MAX=4096 PWM_MAX=32768 STATS_SUMSQ_BASE=256 build_machine_ui
d=May 18 202311:46:31
[INFO] 2024-09-04 20:22:49,413 [root] [temperature_mcu:_mcu_identify:85] mcu_temperature 'mcu' nominal base=378.658537 slope=-804
.878049
[INFO] 2024-09-04 20:22:49,435 [root] [mcu:_send_config:712] Sending MCU 'mcu' printer configuration...
[INFO] 2024-09-04 20:22:49,451 [root] [serialhdl:handle_output:301] mcu 'mcu': #output: allocMax=3804 usedMax=0
[INFO] 2024-09-04 20:22:49,456 [root] [mcu:_connect:785] Configured MCU 'mcu' (3794 moves)
[ERROR] 2024-09-04 20:22:49,463 [root] [klippy:_connect:248] Unhandled exception during connect
Traceback (most recent call last):
  File "/usr/share/klipper/klippy/klippy.py", line 198, in _connect
    cb()
  File "/usr/share/klipper/klippy/mcu.py", line 761, in _connect
    self._send_config(None)
  File "/usr/share/klipper/klippy/mcu.py", line 691, in _send_config
    cb()
  File "/usr/share/klipper/klippy/mcu.py", line 73, in _build_config
    self._trdispatch_mcu = ffi_main.gc(ffi_lib.trdispatch_mcu_alloc(
OverflowError: can't convert negative number to unsigned
`

At the same time machine is barely responsive, sometimes even ssh doesn't work and the most interesting stuff, screen goes crazy, e.g.:

[https://photos.app.goo.gl/VwUPoAcddqYHstbP8]

No clue where to look now, might try to install older version of klipper on Easy BRD but not sure if that might relate with this crazy screen behaviour.

--- a/extras/mmu_toolhead.py
+++ b/extras/mmu_toolhead.py
@@ -36,12 +36,26 @@ class MmuToolHead(toolhead.ToolHead, object):
         self.reactor = self.printer.get_reactor()
         self.all_mcus = [m for n, m in self.printer.lookup_objects(module='mcu')]
         self.mcu = self.all_mcus[0]
+
+        if hasattr(toolhead, 'BUFFER_TIME_HIGH'):
+            time_high = toolhead.BUFFER_TIME_HIGH
+        else:
+            # Backward compatibility for older klipper, like on Sovol or Creality K1 series printers
+            # On Creality K1, these attributes are expected to exist in any Toolhead
+            self.buffer_time_low = config.getfloat('buffer_time_low', 1.000, above=0.)
+            self.buffer_time_high = config.getfloat('buffer_time_high', 2.000, above=self.buffer_time_low)
+            self.buffer_time_start = config.getfloat('buffer_time_start', 0.250, above=0.)
+            self.move_flush_time = config.getfloat('move_flush_time', 0.050, above=0.)
+            self.last_kin_flush_time = self.force_flush_time = self.last_kin_move_time = 0.
+
+            time_high = self.buffer_time_high
+
         if hasattr(toolhead, 'LookAheadQueue'):
             self.lookahead = toolhead.LookAheadQueue(self) # Happy Hare: Use base class LookAheadQueue
-            self.lookahead.set_flush_time(toolhead.BUFFER_TIME_HIGH) # Happy Hare: Use base class
+            self.lookahead.set_flush_time(time_high) # Happy Hare: Use base class
         else:
             self.move_queue = toolhead.MoveQueue(self) # Happy Hare: Use base class MoveQueue (older klipper)
-            self.move_queue.set_flush_time(toolhead.BUFFER_TIME_HIGH) # Happy Hare: Use base class (older klipper)
+            self.move_queue.set_flush_time(time_high) # Happy Hare: Use base class (older klipper)
         self.commanded_pos = [0., 0., 0., 0.]
 
         # MMU velocity and acceleration control
@@ -415,8 +429,10 @@ class MmuKinematics:
             move.limit_speed(self.gear_max_velocity, min(self.gear_max_accel, self.move_accel or self.gear_max_accel))
 
     def get_status(self, eventtime):
+        # Legacy method only used in old Klipper
         return {
             'selector_homed': self.limits[0][0] <= self.limits[0][1],
+            'homed_axes': '' # Required for Creality K1 series printers
         }
 
 

--- a/install.sh
+++ b/install.sh
@@ -3,6 +3,9 @@
 # Installer / Updater script
 #
 # Copyright (C) 2022  moggieuk#6538 (discord) moggieuk@hotmail.com
+# Modified      2024  hamyy <oudy_1999@hotmail.com>
+#               2024  Unsweeticetea <iamzevle@gmail.com>
+#               2024  Dmitry Kychanov <k1-801@mail.ru>
 #
 VERSION=2.70 # Important: Keep synced with mmy.py
 
@@ -12,6 +15,13 @@ SCRIPTPATH="$(dirname "$SCRIPT")"
 SCRIPTNAME="$0"
 ARGS=( "$@" )
 
+# Creality K1 series printers run on MIPS, with a limited instruction set and different default klipper directories
+# Checking for machine type is the easiest way so far to spot them (will be set to 1 if on MIPS):
+IS_MIPS=0
+if [[$(uname -m) = "mips"]]; then
+    IS_MIPS=1
+fi
+
 KLIPPER_HOME="${HOME}/klipper"
 MOONRAKER_HOME="${HOME}/moonraker"
 KLIPPER_CONFIG_HOME="${HOME}/printer_data/config"
@@ -20,6 +30,14 @@ KLIPPER_LOGS_HOME="${HOME}/printer_data/logs"
 OLD_KLIPPER_CONFIG_HOME="${HOME}/klipper_config"
 SENSORS_SECTION="FILAMENT SENSORS"
 LED_SECTION="MMU OPTIONAL NEOPIXEL"
+
+if [ "$IS_MIPS" -eq 1]; then
+    KLIPPER_HOME="/usr/share/klipper"
+    MOONRAKER_HOME="/usr/data/moonraker/moonraker"
+    KLIPPER_CONFIG_HOME="/usr/data/printer_data/config"
+    unset OCTOPRINT_KLIPPER_CONFIG_HOME
+    unset OLD_KLIPPER_CONFIG_HOME
+fi
 
 set -e # Exit immediately on error
 
@@ -193,7 +211,13 @@ self_update() {
     cd "$SCRIPTPATH"
 
     set +e
-    BRANCH=$(timeout 3s git branch --show-current)
+    # timeout is unavailable on MIPS
+    if ["$IS_MIPS" -ne 1]; then
+        BRANCH=$(git branch --show-current)
+    else
+        BRANCH=$(timeout 3s git branch --show-current)
+    fi
+
     if [ $? -ne 0 ]; then
         echo -e "${ERROR}Error updating from github"
         echo -e "${ERROR}You might have an old version of git"
@@ -263,25 +287,44 @@ function nextsuffix {
 }
 
 verify_not_root() {
-    if [ "$EUID" -eq 0 ]; then
-        echo -e "${ERROR}This script must not run as root"
-        exit -1
+    if [ "$IS_MIPS" -ne 1]; then
+        if [ "$EUID" -eq 0 ]; then
+            echo -e "${ERROR}This script must not run as root"
+            exit -1
+        fi
+    else
+        echo -e "${WARNING}This script is running on a MIPS system, so we expect it to be run as root"
     fi
 }
 
 check_klipper() {
     if [ "$NOSERVICE" -ne 1 ]; then
-        if [ "$(sudo systemctl list-units --full -all -t service --no-legend | grep -F "${KLIPPER_SERVICE}")" ]; then
-            echo -e "${INFO}Klipper ${KLIPPER_SERVICE} systemd service found"
+        if [ "$IS_MIPS" -ne 1]; then
+            if [ "$(systemctl list-units --full -all -t service --no-legend | grep -F "${KLIPPER_SERVICE}")" ]; then
+                echo -e "${INFO}Klipper ${KLIPPER_SERVICE} systemd service found"
+            else
+                echo -e "${ERROR}Klipper ${KLIPPER_SERVICE} systemd service not found! Please install Klipper first"
+                exit -1
+            fi
         else
-            echo -e "${ERROR}Klipper ${KLIPPER_SERVICE} systemd service not found! Please install Klipper first"
-            exit -1
+            # There is no systemd on MIPS, we can only check the running processes
+            running_klipper_pid=$(ps -o pid,comm,args | grep [^]]/usr/share/klipper/klippy/klippy.py | awk '{print $1}')
+            KLIPPER_PID_FILE=/var/run/klippy.pid
+
+            if [[ $(cat $KLIPPER_PID_FILE) = $running_klipper_pid ]]; then
+                echo -e "${INFO}Klipper service found"
+            else
+                echo -e "${ERROR}Klipper service not found! Please install Klipper first"
+                exit -1
+            fi
         fi
     fi
 }
 
 check_octoprint() {
-    if [ "$NOSERVICE" -ne 1 ]; then
+    if [ "$IS_MIPS" -eq 1]; then
+        OCTOPRINT=0 # Octoprint can not be set up on MIPS
+    elif [ "$NOSERVICE" -ne 1]; then
         if [ "$(sudo systemctl list-units --full -all -t service --no-legend | grep -F "octoprint.service")" ]; then
             echo -e "${INFO}OctoPrint service found"
             OCTOPRINT=1
@@ -1229,7 +1272,14 @@ uninstall_update_manager() {
 restart_klipper() {
     if [ "$NOSERVICE" -ne 1 ]; then
         echo -e "${INFO}Restarting Klipper..."
-        sudo systemctl restart ${KLIPPER_SERVICE}
+
+        if [ "$IS_MIPS" -ne 1]; then
+            sudo systemctl restart ${KLIPPER_SERVICE}
+        else
+            set +e
+            /etc/init.d/*klipper_service restart
+            set -e
+        fi
     else
         echo -e "${WARNING}Klipper restart suppressed - Please restart ${KLIPPER_SERVICE} by hand"
     fi
@@ -1238,7 +1288,14 @@ restart_klipper() {
 restart_moonraker() {
     if [ "$NOSERVICE" -ne 1 ]; then
         echo -e "${INFO}Restarting Moonraker..."
-        sudo systemctl restart moonraker
+
+        if [ "$IS_MIPS" -ne 1]; then
+            sudo systemctl restart moonraker
+        else
+            set +e
+            /etc/init.d/*moonraker_service restart
+            set -e
+        fi
     else
         echo -e "${WARNING}Moonraker restart suppressed - Please restart by hand"
     fi

